### DISCUSSION TITLE
Sharing SSLClient between MQTT and HTTP?

### DISCUSSION THREAD

**Discussion by marcelstoer** [khoih-prog/EthernetWebServer_SSL#20] [Q&A] [answered]
Do I need a separate `EthernetSSLClient` instance for MQTT and HTTP or can I share it between them?

**Simplified code**

```c++
EthernetClient client;
EthernetSSLClient sslClient(client, TAs, (size_t)TAs_NUM);
PubSubClient mqttClient(MQTT_SERVER, MQTT_PORT, pubSubCallback, sslClient);

// connect to MQTT broker, subscribe to topic in setup()
// feed the mqttClient in loop()

// after new message received through pubSubCallback()
// call sslClient.connect(host, port) to load some resources from an HTTP server

// -> fails with "Arduino client is already connected?"
```

The error is raised from https://github.com/khoih-prog/EthernetWebServer_SSL/blob/main/src/SSLClient/SSLClient_Impl.h#L65. Does it really mean I need two `EthernetSSLClient` instances?

**Comment by khoih-prog** [comment, marked as answer]
I suggest to use multiple instances to avoid future issues.

Check similar issue [**Cannot send requests to different addresses** #4](https://github.com/khoih-prog/AsyncHTTPSRequest_Generic/issues/4), solved by using multiple Client instances.

**Reply by marcelstoer** [reply]
Thanks. It just occurred to me that I could try to increase the SSLClient session cache from 1 to 2 as per https://github.com/OPEnSLab-OSU/SSLClient#session-caching. Will report back.

**Reply by khoih-prog** [reply]
Also try my new [**Ethernet_Generic**](https://github.com/khoih-prog/Ethernet_Generic) library, designed to merge the features of Ethernet, EthernetLarge, Ethernet2, and Ethernet3 libraries. I'll rewrite all my Ethernet-related libs to default using that [**Ethernet_Generic**](https://github.com/khoih-prog/Ethernet_Generic) library.

Feedbacks (bugs, performance, easy-to-use-and-debug, etc.) are very welcome.
